src/script/components/Giphy/index.tsx

```diff
@@ -0,0 +1,307 @@
+import {Gif, GiphyRepository} from '../../extension/GiphyRepository';
+import {getLogger, Logger} from 'Util/Logger';
+
+enum GiphyState {
+  DEFAULT = '',
+  ERROR = 'GiphyViewModel.STATE.ERROR',
+  LOADING = 'GiphyViewModel.STATE.LOADING',
+  NO_SEARCH_RESULT = 'GiphyViewModel.STATE.NO_SEARCH_RESULT',
+  RESULT = 'GiphyViewModel.STATE.RESULT',
+  RESULTS = 'GiphyViewModel.STATE.RESULTS',
+}
+
+interface GiphyProps {
+  readonly giphyRepository: GiphyRepository;
+}
+
+const Giphy: FC<GiphyProps> = ({giphyRepository}) => {
+  const logger: Logger = getLogger('GiphyViewModel');
+
+  const [query, setQuery] = useState<string>('');
+  const [gifs, setGifs] = useState<Gif[]>([]);
+  const [selectedGif, setSelectedGif] = useState<Gif | null>(null);
+  const [currentGif, setCurrentGif] = useState<Gif | null>(null);
+
+  const [giphyState, setGiphyState] = useState<GiphyState>(GiphyState.DEFAULT);
+  const isErrorState = giphyState === GiphyState.ERROR;
+  const isLoadingState = giphyState === GiphyState.LOADING;
+  const isResultState = giphyState === GiphyState.RESULT;
+  const isResultsState = giphyState === GiphyState.RESULTS;
+  const areResultsState = [GiphyState.RESULT, GiphyState.RESULTS].includes(giphyState);
+  const isNoSearchResultState = giphyState === GiphyState.NO_SEARCH_RESULT;
+
+  const hasGifs = gifs.length > 0;
+
+  const loadingTxt = isLoadingState ? t('accessibility.giphyModal.loading') : '';
+
+  const giphyModal = new Modal('#giphy-modal', () => {
+    giphyRepository.resetOffset();
+  });
+
+  const clearGifs = (): void => {
+    setGifs([]);
+    setSelectedGif(null);
+    setGiphyState(GiphyState.LOADING);
+  };
+
+  const getGifs = async (displaySingleResult = false): Promise<void> => {
+    const isStateError = giphyState === GiphyState.ERROR;
+
+    if (isStateError) {
+      return;
+    }
+
+    clearGifs();
+
+    try {
+      const fetchedGifs = await giphyRepository.getGifs(query);
+
+      if (fetchedGifs.length === 0) {
+        setGiphyState(GiphyState.NO_SEARCH_RESULT);
+
+        return;
+      }
+
+      if (fetchedGifs.length === 1 || displaySingleResult) {
+        const gif = fetchedGifs[0];
+
+        setCurrentGif(gif);
+        setGifs([gif]);
+        setSelectedGif(gif);
+        setGiphyState(GiphyState.RESULT);
+      } else {
+        setGifs(fetchedGifs);
+        setGiphyState(GiphyState.RESULTS);
+      }
+    } catch (error) {
+      console.warn(error);
+      setGiphyState(GiphyState.ERROR);
+
+      logger.warn(error);
+    }
+  };
+
+  const onGridClick = async () => getGifs();
+
+  const onBackClick = () => {
+    if (currentGif) {
+      setGifs([currentGif]);
+      setSelectedGif(currentGif);
+    }
+    setGiphyState(GiphyState.RESULT);
+  };
+
+  const onCloseClick = () => {
+    giphyModal.hide();
+    giphyRepository.resetOffset();
+  };
+
+  const showGiphy = async (query: string) => {
+    setQuery(query);
+    setGiphyState(GiphyState.DEFAULT);
+    await getGifs(true);
```

Comment:
⚠️ breaking change: `getGifs` will use currently closed value of `query` (closure), not the one that is being set two lines above (line `128`). We can: 
1. wait for the state to update (`useEffect` with `[query]` deps) and then call `getGifs(true)`, or
2. add query param to `getGifs` function and call it with updated `query` value: `getGifs(query, true)`

I prefer 2nd option 😄